ear/fileio/adm/timing_fixes.py

```diff
@@ -0,0 +1,174 @@
+import warnings
+from ...core.select_items.select_items import _PackAllocator, _ItemSelectionState
+from .elements import AudioBlockFormatObjects, BlockFormat
+
+
+def _has_interpolationLength(blockFormat):
+    """Does blockFormat have a defined interpolationLength?"""
+    assert isinstance(blockFormat, BlockFormat), "wrong type"
+    return (
+        isinstance(blockFormat, AudioBlockFormatObjects)
+        and blockFormat.jumpPosition.flag
+        and blockFormat.jumpPosition.interpolationLength is not None
+    )
+
+
+def fix_blockFormat_durations(adm):
+    """Modify the diration of audioBlockFormats to ensure that the end of one
+    audioBlockFormat always matches the start of the next.
+    """
+    for channelFormat in adm.audioChannelFormats:
+        blockFormat = channelFormat.audioBlockFormats
+
+        for bf_a, bf_b in zip(blockFormat[:-1], blockFormat[1:]):
+            if (
+                bf_a.rtime is None
+                or bf_a.duration is None
+                or bf_b.rtime is None
+                or bf_b.duration is None
+            ):
+                continue
+
+            old_duration = bf_a.duration
+            new_duration = bf_b.rtime - bf_a.rtime
+
+            if old_duration != new_duration:
+                warnings.warn(
+                    "{direction} duration of block format {id}; was: {old}, now: {new}".format(
+                        direction=(
+                            "expanded" if new_duration > old_duration else "contracted"
+                        ),
+                        id=bf_a.id,
+                        old=old_duration,
+                        new=new_duration,
+                    )
+                )
+                bf_a.duration = new_duration
+
+                # if contracting this block makes the interpolation end after
+                # the block, fix it without any more noise
+                if (
+                    _has_interpolationLength(bf_a)
+                    and old_duration >= bf_a.jumpPosition.interpolationLength
+                    and new_duration < bf_a.jumpPosition.interpolationLength
+                ):
+                    bf_a.jumpPosition.interpolationLength = new_duration
+
+
+def fix_blockFormat_interpolationLengths(adm):
+    """Modify the interpolationLength of audioBlockFormats to ensure that they
+    are not greater than the durations.
+    """
+    for channelFormat in adm.audioChannelFormats:
+        for blockFormat in channelFormat.audioBlockFormats:
+            if (
+                blockFormat.rtime is None
+                or blockFormat.duration is None
```

Comment:
This function is just for fixing the interpolation length relative to the block extent (i.e. without considering the object extent). This kind of error (where the interpolation length is greater than the object extent and there is no block timing) should probably be handled too, but isn't currently.

Ideally this case wouldn't occur, because interpolation length in a block with no rtime and duration doesn't have any effect (there can be only one such block in a channel, so it is always the first). I'll add a check for this.

Reply:
Sorry I should have explained that better. What I meant is something like this:

```
<audioBlockFormat audioBlockFormatID="AB_00031001_00000003" rtime="00:00:01.00000">
  <jumpPosition interpolationLength="1.5">1</jumpPosition>
  ...
</audioBlockFormat>
<audioBlockFormat audioBlockFormatID="AB_00031001_00000004" rtime="00:00:02.00000">
  ...
</audioBlockFormat>
```

Reply:
ah, ok. That isn't allowed anyway: https://github.com/ebu/ebu_adm_renderer/blob/master/ear/core/renderer_common.py#L246-L255

Reply:
Ah right, sorry I forgot about that constraint.